native/native_main.nim

```diff
@@ -0,0 +1,244 @@
+            reply.version = some(VERSION)
+
+        of "getconfig":
+            write(stderr, "TODO: NOT IMPLEMENTED\n")
+
+        of "getconfigpath":
+            reply.content = findUserConfigFile()
+            reply.code = some(0)
+            if not isSome(reply.content):
+                reply.code = some(1)
+
+        of "run":
+            # this seems to use /bin/sh rather than the user's shell
+            reply.content = some($ execProcess(msg.command.get(), options={poEvalCommand,poStdErrToStdOut}))
+            reply.code = some 0
+            # probably important to catch the exit code so we can `:cq` in Vim to cancel Ctrl-I
+
+        of "eval":
+            # do we actually want to implement this?
+            # we'd have to start up Python
+            # with whatever stuff is usually used imported
+
+            # should probably deprecate it instead
+            write(stderr, "TODO: NOT IMPLEMENTED\n")
+
+        of "read":
+            try:
+                var f: File
+                discard open(f, expandTilde(msg.file.get()))
+                reply.content = some(readAll(f))
+                reply.code = some(0)
+                close(f)
+            except IOError:
+                reply.content = none(string)
+                reply.code = some(2)
+
+        of "mkdir":
+            try:
+                createDir(expandTilde(msg.dir.get()))
+                reply.content = some("")
+                reply.code = some(0)
+            except OSError:
+                reply.code = some(2)
+
+        of "move":
+            let dest = expandTilde(msg.to.get())
+            if fileExists(dest):
+                reply.code = some 1
+            else:
+                try:
+                    moveFile(dest,msg.`from`.get())
+                    reply.code = some 0
+                except OSError:
+                    reply.code = some 2
+
+        of "write":
+            try:
+                var f: File
+                discard open(f, expandTilde(msg.file.get()), fmWrite)
+                write(f, msg.content.get())
+                reply.code = some(0)
+                close(f)
+            except IOError:
+                reply.code = some(2)
+
+        of "writerc":
+            write(stderr, "TODO: NOT IMPLEMENTED\n")
+
+        of "temp":
+            try:
+                let prefix = "tmp_" & sanitiseFilename(msg.prefix.get("")) & "_"
+                var (f, filepath) = mkstemp(prefix, ".txt", "", fmWrite)
+                write(f, msg.content.get())
+                reply.code = some(0)
+                reply.content = some(filepath)
+                close(f)
+            except IOError:
+                reply.code = some(2)
+
+        of "env":
+            reply.content = some(getEnv(msg.`var`.get()))
+
+        of "list_dir":
+            write(stderr, "TODO: NOT IMPLEMENTED\n")
+
+        of "win_firefox_restart":
+            write(stderr, "TODO: NOT IMPLEMENTED\n")
+
+        of "ppid":
```

Comment:
This causes a compile error when targetting Windows, because it's part of the POSIX module and therefore not available. As far as I can tell, there isn't any good alternative, either, so we'd have to "manually" use the Windows API to make this work on Windows.
It might be best to just platform-gate it and avoid using `ppid` on Windows – as far as I understand it, Windows' process IDs aren't as useful as on *nix anyway.

Reply:
Apparently some of the `posix` module does work on Windows - https://github.com/nim-lang/Nim/issues/9372#issuecomment-564350044 - but yeah, we can just gate it since it's only used by `ff_cmdline` if the platform isn't Windows.

Reply:
We should be able to gate it with something like this:
```nim
        of "ppid":
            when defined posix:
                reply.content = some($getppid())
            else:
                reply.cmd = some("error")
                reply.error = some("ppid is not available on this operating system")
```
Or we could just have it be handled by the else at the end and treat it as any other unknown command:
```nim
        of (when defined posix: "ppid" else: {}):
            reply.content = some($getppid())
```

Reply:
I think it might need to be `when defined getppid` because some POSIX stuff is available on Windows like `mkstemp`.

Reply:
`defined posix` doesn't check whether `posix` has been imported, it checks whether a compile-time "external symbol" called "posix" has been defined, which the compiler does automatically when targeting POSIX systems. What you're thinking of is called [`declared`](https://nim-lang.org/docs/system.html#declared%2Cuntyped) in Nim, but I've tried `when declared getppid` already, but that always returns true, regardless of whether it actually works on the platform you're targeting.